### PR TITLE
fix(profiles): RHICOMPL-2085 sort profiles by policy name if set

### PR DESCRIPTION
**Pull request by skateman**
I wanted to avoid using the `default_scope` and go with the `ProfilesController#resource` method having the `left_outer_joins` call. However, I did not know how to do the same thing in GraphQL :thinking: :disappointed: if anyone knows, please let me know in the review :pray:

### CODE REVIEW

File: app/models/profile.rb [hunk 1]
```diff
@@ -11,11 +11,13 @@ class Profile < ApplicationRecord
   include ProfileRules
 
   SORTABLE_BY = {
-    name: :name,
+    name: Arel.sql('COALESCE(policies.name, profiles.name)'),
     os_minor_version: :os_minor_version,
     score: :score
   }.freeze
 
+  default_scope { left_outer_joins(:policy) }
```
Comment (vkrizan) [inline app/models/profile.rb:19]: I wonder whether this would not cause any issues. Like an ambiguous column select.

**Review by vkrizan — COMMENTED** [review]
For GraphQL check out https://github.com/RedHatInsights/compliance-backend/blob/master/app/graphql/resolvers/concerns/collection.rb#L26

I'd also like this to be controller specific.

File: app/models/profile.rb (hunk 1 above)
Comment (skateman) [reply app/models/profile.rb:19]: That only happens when you're using string-based attributes in `ActiveRecord::Base#select`, passing a symbol to this method would always default to `tablename.field`. You can do a full text search for `.select(` and find all of these easily.

**Comment by skateman** [comment]
I don't see how I can make the `authorized_scope˙ type-specific :thinking:

**Review by akofink — APPROVED** [review]
Works great, thanks! This default_scope seems OK for now, since it's a left outer join and shouldn't change the ultimate results of the query; however, it could affect performance. Most of the places we use profile require a policy anyway, but the default scope affects _every_ place we use `Profile`, not just within our controllers (i.e. migrations, too). In places where we definitely don't want the default scope, we should be using `Profile.unscope...`. Eventually, we should extend the sorting code to allow additional includes/joins.

**Comment by skateman** [comment]
> Eventually, we should extend the sorting code to allow additional includes/joins.

This is what I needed, thank you @akofink :pray: :tada: :1st_place_medal: :sweat_smile: I remade the `SORTABLE_BY` into a simple DSL method that takes the column, the SQL statement and optionally the additional scope as arguments:
```ruby
sortable_by :name, Arel.sql('COALESCE(policies.name, profiles.name)'), scope: :joins_policy
```

Code is coming in a second, last set of tests are running :turtle: 

PS: you might not like the idea of moving a bigger part of the sort-by building into the models, but I guess we can negotiate on how much should be where :thinking:

**Review by akofink — APPROVED** [review]
Awesome job, @skateman! Nothing major - it works as expected, and I love the new DSL.

File: app/services/duplicate_canonical_profile_resolver.rb [hunk 2]
```diff
@@ -23,12 +23,12 @@ def duplicates
...
         :benchmark_id,
         :policy_id,
-        'COUNT(id)'
+        'COUNT(profiles.id)'
       ).group(
         :ref_id,
         :benchmark_id,
         :policy_id
-      ).having('count(id) > 1')
+      ).having('count(profiles.id) > 1')
```
[the remark this answers is not in the document]
Comment (skateman) [reply app/services/duplicate_canonical_profile_resolver.rb:31]: Missed it, fixing :wrench:

File: app/models/concerns/sortable.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    private
+
+    def sortable_by(column, statement, scope: nil)
+      @sortable_by ||= {}
+      @sortable_by[column] = {
+        statement: statement,
+        scope: scope
+      }
+    end
```
[the remark this answers is not in the document]
Comment (skateman) [reply app/models/concerns/sortable.rb:26]: I just realized that this is valid in Ruby:
```ruby
def sortable_by(column, statement = column, scope: nil)
```
:heart_eyes:

**Review by akofink — APPROVED** [review]

**Pull request merged**